statistics/row_sampler.go

```diff
@@ -0,0 +1,293 @@
+}
+
+// RowSampleBuilder is used to construct the RowSampleCollector to get the samples.
+type RowSampleBuilder struct {
+	Sc              *stmtctx.StatementContext
+	RecordSet       sqlexec.RecordSet
+	ColsFieldType   []*types.FieldType
+	Collators       []collate.Collator
+	ColGroups       [][]int64
+	MaxSampleSize   int
+	MaxFMSketchSize int
+	Rng             *rand.Rand
+}
+
+// Collect first builds the collector. Then maintain the null count, FM sketch and the data size for each column and
+// column group.
+// Then use the weighted reservoir sampling to collect the samples.
+func (s *RowSampleBuilder) Collect() (*RowSampleCollector, error) {
+	collector := &RowSampleCollector{
+		Samples:       make(WeightedRowSampleHeap, 0, s.MaxSampleSize),
+		NullCount:     make([]int64, len(s.ColsFieldType)+len(s.ColGroups)),
+		FMSketches:    make([]*FMSketch, 0, len(s.ColsFieldType)+len(s.ColGroups)),
+		TotalSizes:    make([]int64, len(s.ColsFieldType)+len(s.ColGroups)),
+		MaxSampleSize: s.MaxSampleSize,
+	}
+	for i := 0; i < len(s.ColsFieldType)+len(s.ColGroups); i++ {
+		collector.FMSketches = append(collector.FMSketches, NewFMSketch(s.MaxFMSketchSize))
+	}
+	ctx := context.TODO()
+	chk := s.RecordSet.NewChunk()
+	it := chunk.NewIterator4Chunk(chk)
+	for {
+		err := s.RecordSet.Next(ctx, chk)
+		if err != nil {
+			return nil, err
+		}
+		if chk.NumRows() == 0 {
+			return collector, nil
+		}
+		collector.Count += int64(chk.NumRows())
+		for row := it.Begin(); row != it.End(); row = it.Next() {
+			datums := RowToDatums(row, s.RecordSet.Fields())
+			for i, val := range datums {
+				// For string values, we use the collation key instead of the original value.
+				if s.Collators[i] != nil && !val.IsNull() {
+					decodedVal, err := tablecodec.DecodeColumnValue(val.GetBytes(), s.ColsFieldType[i], s.Sc.TimeZone)
+					if err != nil {
+						return nil, err
+					}
+					decodedVal.SetBytesAsString(s.Collators[i].Key(decodedVal.GetString()), decodedVal.Collation(), uint32(decodedVal.Length()))
+					encodedKey, err := tablecodec.EncodeValue(s.Sc, nil, decodedVal)
+					if err != nil {
+						return nil, err
+					}
+					val.SetBytes(encodedKey)
+				}
+				err := collector.collectColumns(s.Sc, datums)
+				if err != nil {
+					return nil, err
+				}
+				err = collector.collectIndexes(s.Sc, datums, s.ColGroups)
+				if err != nil {
+					return nil, err
+				}
+			}
+			weight := s.Rng.Int63()
+			collector.sampleRow(datums, weight)
+		}
+	}
+}
+
+func (s *RowSampleCollector) collectColumns(sc *stmtctx.StatementContext, cols []types.Datum) error {
+	for i, col := range cols {
+		if col.IsNull() {
+			s.NullCount[i]++
+		}
+		s.TotalSizes[i] += int64(len(col.GetBytes())) - 1
+		// Minus one is to remove the flag byte.
+		err := s.FMSketches[i].InsertValue(sc, col)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *RowSampleCollector) collectIndexes(sc *stmtctx.StatementContext, cols []types.Datum, colGroups [][]int64) error {
```

Comment:
How about renaming it `collectColGroups` since we want to get rid of the "index" concept in analyze related code in the future.